State only modifies Context#state - prep for Map/Parallel

kbrock opened the pull request:
In prep for `Map`, have State only modify its own data in `Context`. (Followup will pass `context` into `State` methods, so the caller (i.e.: `Map`, `Parallel`, or `Workspace`) will determine the context and act accordingly.

1. Don't modify the context in `Workflow.new` - instead modify it in the first step.
2. `State#start` only modifies `Context#state` and not `Context#execute` - instead `Workflow` is responsible for `execute`.
3. `State#finish` only modified `Context#state` and not `Context#execute` - instead `Workflow` is responsible for that.

kbrock commented:
update:
- fixed cops
- added comment
- indentation

agrare commented on lib/floe/workflow.rb:
```diff
@@ -124,6 +119,12 @@ def step_nonblock
 
       step_next
       current_state.run_nonblock!
+                   .tap do
+        # Avoiding State#running? because that is potentially expensive.
+        # State#run_nonblock! already called running? via State#ready? and
+        # called State#finished -- which is what Context#state_finished? is detecting
+        context.execution["EndTime"] = context.state["FinishedTime"] if context.state_finished? && context.next_state.nil?
```
The .tap without actually using the return looks odd, could we just do
```ruby
result = current_state.run_nonblock!
context.execution["EndTime"] = context.state["FinishedTime"] if context.state_finished? && context.next_state.nil?
result
```

agrare replied:
And if we have a `start_workflow!` [[Ref]](https://github.com/ManageIQ/floe/pull/206/files#r1631302239) we could call this `end_workflow!` or `finish_workflow!`

kbrock replied:
really like this

agrare commented on lib/floe/workflow.rb:
```diff
@@ -155,6 +156,11 @@ def end?
     end
 
     def current_state
+      unless context.state.key?("Name")
+        context.state["Name"] = start_at
+        context.state["Input"] = context.execution["Input"].dup
+        context.execution["StartTime"] = Time.now.utc.iso8601
+      end
```
I think asking for the current_state potentially setting it is a bit overloaded, what if we had a `start_workflow!` method which set the workflow execution["StartTime"] and set up the initial state?  Then just either handle if `current_state.nil?` or call `start_workflow!` before any checks to `current_state`

kbrock replied:
made these changes.
you can see that `current_state` which is called a lot in specs needed the `start_workflow` - so a bunch of changes

I created a commit before these changes that reworked tests to have less stubbing and possibly call current_state less. This reduced in the test changes.

kbrock commented:
update:
- created method for start_workflow
- moved start_workflow out of current_state - this caused a lot of spec churn

update:
- oops, added end_workflow
- cops

kbrock commented:
@agrare looking forward to seeing #208 leveraging this
